src/object/srv_obj.c

```diff
@@ -790,129 +797,152 @@ ds_obj_check_dtx_config(uint32_t flags, struct dtx_id *dti)
-				&cont_hdl, &cont);
+	rc = ds_pre_check(orw->orw_map_ver, orw->orw_co_hdl, orw->orw_co_uuid,
+			  opc_get(rpc->cr_opc), &cont_hdl, &cont);
 	if (rc)
 		goto out;
 
-	if (update && !(cont_hdl->sch_capas & DAOS_COO_RW)) {
-		D_ERROR("cont "DF_UUID" sch_capas "DF_U64", "
-			"NO_PERM to update.\n",
-			DP_UUID(orw->orw_co_uuid), cont_hdl->sch_capas);
-		D_GOTO(out, rc = -DER_NO_PERM);
-	}
-
 	D_ASSERT(cont_hdl->sch_pool != NULL);
 	map_ver = cont_hdl->sch_pool->spc_map_version;
 
 	D_DEBUG(DB_TRACE, "rpc %p opc %d "DF_UOID" dkey %d %s tag/xs %d/%d eph "
 		DF_U64", pool ver %u/%u with "DF_DTI".\n",
 		rpc, opc_get(rpc->cr_opc), DP_UOID(orw->orw_oid),
 		(int)orw->orw_dkey.iov_len, (char *)orw->orw_dkey.iov_buf,
-		tag, dss_get_module_info()->dmi_xs_id, orw->orw_epoch,
+		dss_get_module_info()->dmi_tgt_id,
+		dss_get_module_info()->dmi_xs_id, orw->orw_epoch,
 		orw->orw_map_ver, map_ver, DP_DTI(&orw->orw_dti));
 
-	if (orw->orw_map_ver < map_ver) {
-		D_DEBUG(DB_IO, "stale version req %d map_version %d\n",
-			orw->orw_map_ver, map_ver);
-		if (update)
-			D_GOTO(out, rc = -DER_STALE);
-
-		/* It is harmless if fetch with old pool map version. */
-	} else if (orw->orw_map_ver > map_ver) {
-		/* XXX: Client (or leader replica) has newer pool map than
-		 *	current replica. Two possibile cases:
-		 *
-		 *	1. The current replica was the old leader if with
-		 *	   the old pool map version. According to current
-		 *	   leader election algorithm, it is still the new
-		 *	   leader with the new pool map version. Since no
-		 *	   leader switch, the unmatched pool version will
-		 *	   not affect DTX related availability check.
-		 *
-		 *	2. The current replica was NOT the old leader if
-		 *	   with the old pool map version. But it becomes
-		 *	   the new leader with the new pool map version.
-		 *	   In the subsequent modificaiton, it may hit
-		 *	   some 'prepared' DTX when make availability
-		 *	   check, it will return -DER_INPROGRESS that
-		 *	   will cause client to retry. It is possible
-		 *	   that the pool map version event arrives at
-		 *	   this server during the client retry. It is
-		 *	   inefficient, but harmless.
-		 */
-	}
-
-	if (!update) {
-		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_PREP);
+	/** Three cases
+	 * 1. Single replica update.
+	 * 2. Non-leader update.
+	 * 3. Fetch.
+	 */
+	if (orw->orw_shard_tgts.ca_arrays == NULL) {
+		/* No need resend check? XXX */
 		D_TIME_START(tls->ot_sp, OBJ_PF_UPDATE_LOCAL);
-
-		rc = ds_obj_rw_local_hdlr(rpc, tag, cont_hdl, cont,
-					  &ioh, false, map_ver, dth);
+		rc = ds_obj_rw_local_hdlr(rpc, cont_hdl, cont,
+					  &ioh, map_ver, dth);
 		if (rc != 0)
-			D_ERROR(DF_UOID": rw_local_hdlr (fetch) failed %d.\n",
+			D_ERROR(DF_UOID": rw_local_hdlr (update) failed %d.\n",
 				DP_UOID(orw->orw_oid), rc);
 
 		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_LOCAL);
-
 		D_GOTO(out, rc);
 	}
 
-	if (!daos_oc_echo_type(daos_obj_id2class(orw->orw_oid.id_pub))) {
```

Comment:
Why remove the eoch case handling?

Reply:
Hmm, I thought echo is suppose to "measure" the overhead before VOS, so it is no harm to let echo to go through this? or I miss sth?

Reply:
The echo mode will not update the on-disk data, then resend check will always invalid.